modules/page-memory/src/test/java/org/apache/ignite/internal/pagememory/persistence/PageMemoryImplNoLoadTest.java

```diff
@@ -78,56 +82,193 @@ public void testPageHandleDeallocation() {
+        dataRegionCfg.change(c -> c.changeInitSize(128 * systemPageSize).changeMaxSize(128 * systemPageSize)).get(1, SECONDS);
 
-        PageMemoryImpl memory = memory(new long[]{100 * systemPageSize, 28 * systemPageSize});
+        PageMemoryImpl pageMemoryImpl = createPageMemoryImpl(
+                new long[]{100 * systemPageSize, 28 * systemPageSize},
+                filePageStoreManager,
+                checkpointManager
+        );
+
+        dataRegions.add(newDataRegion(true, pageMemoryImpl));
+
+        filePageStoreManager.start();
+
+        checkpointManager.start();
 
-        memory.start();
+        pageMemoryImpl.start();
 
         try {
-            long maxPages = memory.totalPages();
+            initGroupFilePageStores(filePageStoreManager);
+
+            long maxPages = pageMemoryImpl.totalPages();
 
             long maxDirtyPages = (maxPages * 3 / 4);
 
             assertThat(maxDirtyPages, greaterThanOrEqualTo(50L));
 
-            for (int i = 0; i < maxDirtyPages - 1; i++) {
-                allocatePage(memory);
+            checkpointManager.checkpointTimeoutLock().checkpointReadLock();
 
-                assertTrue(memory.safeToUpdate(), "i=" + i);
-            }
+            try {
+                for (int i = 0; i < maxDirtyPages - 1; i++) {
+                    createDirtyPage(pageMemoryImpl);
 
-            for (int i = (int) maxDirtyPages - 1; i < maxPages; i++) {
-                allocatePage(memory);
+                    assertTrue(pageMemoryImpl.safeToUpdate(), "i=" + i);
+                }
 
-                assertFalse(memory.safeToUpdate(), "i=" + i);
+                for (int i = (int) maxDirtyPages - 1; i < maxPages; i++) {
+                    createDirtyPage(pageMemoryImpl);
+
+                    assertFalse(pageMemoryImpl.safeToUpdate(), "i=" + i);
+                }
+            } finally {
+                checkpointManager.checkpointTimeoutLock().checkpointReadUnlock();
             }
 
-            // TODO: IGNITE-16984 After the checkpoint check assertTrue(memory.safeToUpdate())
+            checkpointManager
+                    .forceCheckpoint("for_test_safe_to_update", null)
+                    .futureFor(FINISHED)
+                    .get(100, MILLISECONDS);
+
+            assertTrue(pageMemoryImpl.safeToUpdate());
         } finally {
-            memory.stop(true);
+            closeAll(
+                    () -> pageMemoryImpl.stop(true),
+                    checkpointManager::stop,
+                    filePageStoreManager::stop
+            );
+        }
+    }
+
+    protected PageMemoryImpl createPageMemoryImpl(
+            long[] sizes,
+            @Nullable FilePageStoreManager filePageStoreManager,
+            @Nullable CheckpointManager checkpointManager
+    ) {
+        PageIoRegistry ioRegistry = new PageIoRegistry();
+
+        ioRegistry.loadFromServiceLoader();
+
+        return new PageMemoryImpl(
+                new UnsafeMemoryProvider(null),
```

Comment:
Why is this a parameter? I thought we should derive this instance from the configuration. There's a specific config value for allocator type.

Reply:
Then we don't need interface **DirectMemoryProvider** and will always be **UnsafeMemoryProvider** ?

Reply:
No, there's NUMA allocator coming!

Reply:
I don't know then, design is taken from 2.x codebase... Is this the only planned implementation? Can you check implementations in 2.x?

Reply:
There is also this one: **org.apache.ignite.internal.mem.file.MappedFileMemoryProvider**
Since this is a test, I suggest leaving it as is.

Reply:
Fix it